FAILED: ActorableExampleRouterTests test_actorRouter_shouldWork

Reported via nio release tools.

Failure on macOS, thanks for reporting @PeterAdams-A 


```
Test Case ‘-[DistributedActorsTests.ActorableExampleRouterTests test_actorRouter_shouldWork]’ started.
/private/tmp/.nio-release-tools_tH3oVA/swift-distributed-actors/Sources/DistributedActorsTestKit/TestProbes.swift:447: error: -[DistributedActorsTests.ActorableExampleRouterTests te
st_actorRouter_shouldWork] : failed: caught error: ”
    try p.expectMessagesInAnyOrder([
       ESC[0;31m^~~~~~~~~~~~~~~~~~~~~~~~~
error: Received only (in order) [[“one: foo”, “one: bar”]], but did not receive expected [[“two: foo”, “two: bar”]]:Swift.String within [5s], error: noMessagesInQueueESC[0;0m”
[pprint][9:5:50.8730] [/private/tmp/.nio-release-tools_tH3oVA/swift-distributed-actors/Tests/DistributedActorsTests/Actorable/ActorableExampleRouterTests.swift:271][thread:1359651]:
 key = Reception.Key<ActorRef<Message>>(id: *)
------------------------------------- ActorSystem(ActorableExampleRouterTests) ------------------------------------------------
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [ActorSystem.swift:307] [info] Actor System [ActorableExampleRouterTests] initialized.
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [OperationLogClusterReceptionist.swift:229][/system/receptionist] [debug] Initialized receptionist
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [ClusterEventStream.swift:43][/system/clusterEvents] [trace] Successfully subscribed [ActorRef<Cluster.Event>(/sys
tem/replicator/$sub-DistributedActors.Cluster.Event-y)], offering membership snapshot
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [ClusterEventStream.swift:43][/system/clusterEvents] [trace] Successfully subscribed [ActorRef<Cluster.Event>(/sys
tem/receptionist/$sub-DistributedActors.Cluster.Event-y)], offering membership snapshot
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [Gossiper+Shell.swift:334][/system/replicator/gossip] [debug] Registered with receptionist key: Reception.Key<Acto
rRef<Message>>(id: crdt-gossip-replicator)
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [OperationLogClusterReceptionist.swift:339][/system/receptionist] [debug] Registered [/system/replicator/gossip] f
or key [AnyReceptionistKey<DistributedActors.ActorRef<DistributedActors.GossipShell<DistributedActors.CRDT.Gossip, DistributedActors.CRDT.GossipAck>.Message>>(crdt-gossip-replicator
)]
// metadata:
// “receptionist/key”: AnyReceptionistKey<DistributedActors.ActorRef<DistributedActors.GossipShell<DistributedActors.CRDT.Gossip, DistributedActors.CRDT.GossipAck>.Message>>(crdt-go
ssip-replicator)
// “receptionist/opLog/maxSeqNr”: 1
// “receptionist/registered”: /system/replicator/gossip
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [OperationLogClusterReceptionist.swift:305][/system/receptionist] [trace] Subscribed /system/replicator/gossip/$su
b-DistributedActors.Reception.Listing<DistributedActors.ActorRef<DistributedActors.GossipShell<DistributedActors.CRDT.Gossip,DistributedActors.CRDT.GossipAck>.Message>>-y to AnyRece
ptionistKey<DistributedActors.ActorRef<DistributedActors.GossipShell<DistributedActors.CRDT.Gossip, DistributedActors.CRDT.GossipAck>.Message>>(crdt-gossip-replicator)
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8720] [Gossiper+Shell.swift:337][/system/replicator/gossip] [trace] Peer listing update via receptionist
// metadata:
// “peer/listing”: 
//  ActorRef<GossipShell<DistributedActors.CRDT.Gossip, DistributedActors.CRDT.GossipAck>.Message>(/system/replicator/gossip)
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8730] [OperationLogClusterReceptionist.swift:339][/system/receptionist] [debug] Registered [/user/one] for key [AnyRecep
tionistKey<DistributedActors.ActorRef<DistributedActorsTests.BarActorTarget.Message>>(*)]
// metadata:
// “receptionist/key”: AnyReceptionistKey<DistributedActors.ActorRef<DistributedActorsTests.BarActorTarget.Message>>(*)
// “receptionist/opLog/maxSeqNr”: 2
// “receptionist/registered”: /user/one
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:50.8730] [OperationLogClusterReceptionist.swift:339][/system/receptionist] [debug] Registered [/user/two] for key [AnyRecep
tionistKey<DistributedActors.ActorRef<DistributedActorsTests.FooActorTarget.Message>>(*)]
// metadata:
// “receptionist/key”: AnyReceptionistKey<DistributedActors.ActorRef<DistributedActorsTests.FooActorTarget.Message>>(*)
// “receptionist/opLog/maxSeqNr”: 3
// “receptionist/registered”: /user/two
Captured log [ActorableExampleRouterTests][2020-07-28 9:5:51.1260] [OperationLogClusterReceptionist.swift:381][/system/receptionist] [trace] Delayed listing flush: AnyReceptionTest 
Case ‘-[DistributedActorsTests.ActorableExampleRouterTests test_actorRouter_shouldWork]’ failed (7.099 seconds).
```

--- a/Tests/DistributedActorsTests/Actorable/ActorableExampleRouterTests.swift
+++ b/Tests/DistributedActorsTests/Actorable/ActorableExampleRouterTests.swift
@@ -16,9 +16,6 @@
 import DistributedActorsTestKit
 import XCTest
 
-// ==== ----------------------------------------------------------------------------------------------------------------
-// MARK: Tests
-
 final class ActorableExampleRouterTests: ActorSystemXCTestCase {
     func test_actorRouter_viaReceptionist_shouldWork() throws {
         let router: Actor<MessageActorReceptionistRouter> = try system.spawn("ref-router") { _ in
@@ -120,6 +117,7 @@ final class MessageActorReceptionistRouter: Actorable {
         self.listing = context.receptionist.autoUpdatedListing(key)
 
         self.listing.onUpdate { _ in
+            // attempt to deliver messages (or buffer them again, this is an inefficient way to implement this, but good enough for our test)
             let buffered = self.buffer
             self.buffer = []
 
@@ -147,9 +145,19 @@ final class MessageActorReceptionistRouter: Actorable {
 final class MessageRefRouter: Actorable {
     var targets: [String: ActorRef<GeneratedActor.Messages.MessageTarget>] = [:]
 
+    var buffer: [(String, MessageType, ActorRef<String>)] = []
+
     /* @actor */
     func registerTarget(_ identifier: String, actor ref: ActorRef<GeneratedActor.Messages.MessageTarget>) {
         self.targets[identifier] = ref
+
+        // attempt to deliver messages (or buffer them again, this is an inefficient way to implement this, but good enough for our test)
+        let buffered = self.buffer
+        self.buffer = []
+
+        buffered.forEach { identifier, message, replyTo in
+            self.dispatchMessage(identifier, message: message, replyTo: replyTo)
+        }
     }
 
     /* @actor */
@@ -166,6 +174,8 @@ final class MessageRouter: Actorable {
     // Map of target identifier to Actor
     var targets: [String: AnyActor.MessageTarget]
 
+    var buffer: [(String, MessageType, ActorRef<String>)] = []
+
     init() {
         self.targets = [:]
     }
@@ -174,6 +184,14 @@ final class MessageRouter: Actorable {
     /* @actor */
     func registerTarget(_ identifier: String, actor: AnyActor.MessageTarget) {
         self.targets[identifier] = actor
+
+        // attempt to deliver messages (or buffer them again, this is an inefficient way to implement this, but good enough for our test)
+        let buffered = self.buffer
+        self.buffer = []
+
+        buffered.forEach { identifier, message, replyTo in
+            self.dispatchMessage(identifier, message: message, replyTo: replyTo)
+        }
     }
 
     // Invoked by some other component receiving messages from an external
@@ -267,8 +285,7 @@ struct BarActorTarget: Actorable, MessageTarget {
         )
 
         // also register with receptionist
-        let key = context.receptionist.register(context.myself.ref, as: "*")
-        pprint("key = \(key)")
+        let _: Reception.Key<ActorRef<BarActorTarget.Message>> = context.receptionist.register(context.myself.ref, as: "*")
     }
 
     // @actor

